rpcs3/Emu/RSX/VK/VKOverlays.cpp

```diff
@@ -887,50 +887,81 @@ namespace vk
 			"}\n";
 
 		fs_src =
-			"#version 420\n\n"
-			"layout(set=0, binding=1) uniform sampler2D fs0;\n"
-			"layout(set=0, binding=2) uniform sampler2D fs1;\n"
-			"layout(location=0) in vec2 tc0;\n"
-			"layout(location=0) out vec4 ocol;\n"
-			"\n"
-			"layout(push_constant) uniform static_data\n"
-			"{\n"
-			"	float gamma;\n"
-			"	int limit_range;\n"
-			"	int stereo;\n"
-			"	int stereo_image_count;\n"
-			"};\n"
-			"\n"
-			"vec4 read_source()\n"
-			"{\n"
-			"	if (stereo == 0) return texture(fs0, tc0);\n"
-			"\n"
-			"	vec4 left, right;\n"
-			"	if (stereo_image_count == 2)\n"
-			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
-			"	}\n"
-			"	else\n"
-			"	{\n"
-			"		vec2 coord_left = tc0 * vec2(1.f, 0.4898f);\n"
-			"		vec2 coord_right = coord_left + vec2(0.f, 0.510204f);\n"
-			"		left = texture(fs0, coord_left);\n"
-			"		right = texture(fs0, coord_right);\n"
-			"	}\n"
-			"\n"
-			"	return vec4(left.r, right.g, right.b, 1.);\n"
-			"}\n"
-			"\n"
-			"void main()\n"
-			"{\n"
-			"	vec4 color = read_source();\n"
-			"	color.rgb = pow(color.rgb, vec3(gamma));\n"
-			"	if (limit_range > 0)\n"
-			"		ocol = ((color * 220.) + 16.) / 255.;\n"
-			"	else\n"
-			"		ocol = color;\n"
-			"}\n";
+      "#version 420\n\n"
+      "layout(set=0, binding=1) uniform sampler2D fs0;\n"
+      "layout(set=0, binding=2) uniform sampler2D fs1;\n"
+      "layout(location=0) in vec2 tc0;\n"
+      "layout(location=0) out vec4 ocol;\n"
+      "\n"
+      "layout(push_constant) uniform static_data\n"
+      "{\n"
+      "	float gamma;\n"
+      "	int limit_range;\n"
+      "	int stereo;\n"
+      "	int stereo_image_count;\n"
+      "	int stereo_display_method;\n"
+      "};\n"
+      "\n"
+      "vec4 read_source()\n"
+      "{\n"
+      "	if (stereo == 0) return texture(fs0, tc0);\n"
+      "\n"
+      "	vec4 left, right;\n"
+      "	if (stereo_image_count == 2)\n"
```

Comment:
This whole section is wrong.
You can get 1 image or 2 with 3D. This is an internal detail.
PS3 renders over-under, but some games will write both eyes in one pass (so we have one texture cached) and some will have 2 passes (so we have 2 textures)
The read source method should not have been altered at all, the only thing you should change is how you write the outputs. Generate the inputs at the vertex stage, not the fragment stage.
i.e If you want SBS/Over-under, you should emit 2 draw calls, one for each eye.
The only reason this seems to have even worked seems to have been by chance, you were testing a game that emits 2 draw calls and got 2 separate images.

Reply:
I suggest to use multi-draw (instanced) and use shader_draw_parameters to pick which coordinates to emit.

To summarize:
- This fragment output shader should really have no logic to handle any of these, you can send those to the VS and place them correctly.
- When doing SBS/Over-Under you no longer have stereo output, so you need a new variable to only read source but composite for one eye only.
- VS should just have a set of coordinates as constants and just emit the relevant ones for the current mode and eye.

Reply:
ah, could you name a game that does the other method? i was testing with killzone 3.

also i'm not quite sure how i'd do that at the vertex stage, i could probably figure it out but having slightly more info when i'm able to work on it again wouldn't hurt.

Reply:
Try motorstorm and pacific rift, it's been a long time, I cannot remember which game used which mode.

If doing it with verts is too hard, you can still do it with the fragment shader, though the logic exhibited in this PR is incorrect. The easiest way would be to use a compute shader with image_load_store instead of draw calls. The logic that cares about the stereo mode should be after the read_source() call where you already have a left and right color. You would then use a 2D matrix to place the fragment in its correct eye coordinate. Luckily, as we don't do warping or rotation, your matrix is a simple scale and translate matrix that should take fs0 in and give you a left and right coord.

To do it with a draw call pass is a reverse operation. Instead of reading fs0, you would do an inverse projection with the 2D matrix to get the original eye id (left/right) and the coordinate within the eye. You would then call read_source, select the color you want (left/right) and write to fragment out as usual.

Reply:
i'm not calculating it, it doesn't matter as it's not VR.

the VR viewing app i used displayed it fine, as a floating screen in front of me, with one half of the window in each eye respectively

![Screenshot_20230326_022836](https://user-images.githubusercontent.com/16214950/228022418-795addb8-a9d6-4b21-b908-0c08bd9d48c4.png)